src/legacy/shrink.rs

```diff
@@ -0,0 +1,379 @@
+    /// the queue is empty.
+    fn remove_next(&mut self) -> Option<u16> {
+        let res = self.next();
+        if res.is_some() {
+            self.next_idx += 1;
+        }
+        res
+    }
+}
+
+#[derive(Clone, Debug, Copy)]
+struct Codetab {
+    prefix_code: Option<u16>,
+    ext_byte: u8,
+    len: u16,
+    last_dst_pos: usize,
+}
+
+impl Default for Codetab {
+    fn default() -> Self {
+        Self {
+            prefix_code: None,
+            ext_byte: 0,
+            len: 0,
+            last_dst_pos: 0,
+        }
+    }
+}
+
+impl Codetab {
+    pub fn create_new() -> [Self; MAX_CODE + 1] {
+        let mut codetab = (0..=u8::MAX)
+            .map(|i| Codetab {
+                prefix_code: Some(i as u16),
+                ext_byte: i,
+                len: 1,
+                last_dst_pos: 0,
+            })
+            .collect::<Vec<_>>();
+        codetab.resize(MAX_CODE + 1, Codetab::default());
+        codetab.try_into().unwrap()
+    }
+}
+
+fn unshrink_partial_clear(codetab: &mut [Codetab], queue: &mut CodeQueue) {
+    let mut is_prefix = [false; MAX_CODE + 1];
+
+    // Scan for codes that have been used as a prefix.
+    for i in CONTROL_CODE + 1..=MAX_CODE {
+        if let Some(prefix_code) = codetab[i].prefix_code {
+            is_prefix[prefix_code as usize] = true;
+        }
+    }
+
+    // Clear "non-prefix" codes in the table; populate the code queue.
+    let mut code_queue_size = 0;
+    for i in CONTROL_CODE + 1..MAX_CODE {
+        if !is_prefix[i] {
+            codetab[i].prefix_code = None;
+            queue.codes[code_queue_size] = Some(i as u16);
+            code_queue_size += 1;
+        }
+    }
+    queue.codes[code_queue_size] = None; // End-of-queue marker.
+    queue.next_idx = 0;
+}
+
+/// Read the next code from the input stream and return it in next_code. Returns
+/// false if the end of the stream is reached. If the stream contains invalid
+/// data, next_code is set to INVALID_CODE but the return value is still true.
+fn read_code<T: std::io::Read, E: Endianness>(
+    is: &mut BitReader<T, E>,
+    code_size: &mut u8,
+    codetab: &mut [Codetab],
+    queue: &mut CodeQueue,
+) -> io::Result<Option<u16>> {
+    // assert(sizeof(code) * CHAR_BIT >= *code_size);
+    let code = is.read::<u16>(*code_size as u32)?;
+
+    // Handle regular codes (the common case).
+    if code != CONTROL_CODE as u16 {
+        return Ok(Some(code));
+    }
+
+    // Handle control codes.
+    let control_code = if let Ok(c) = is.read::<u16>(*code_size as u32) {
+        c
+    } else {
+        return Ok(None);
+    };
+    if control_code == INC_CODE_SIZE && *code_size < MAX_CODE_SIZE {
+        (*code_size) += 1;
+        return read_code(is, code_size, codetab, queue);
+    }
+    if control_code == PARTIAL_CLEAR {
+        unshrink_partial_clear(codetab, queue);
+        return read_code(is, code_size, codetab, queue);
+    }
+    return Ok(None);
+}
```

Comment:
We can rule out `Ok(None)` by handling a few cases of invalid input.
```suggestion
fn read_code<T: std::io::Read, E: Endianness>(
    is: &mut BitReader<T, E>,
    code_size: &mut u8,
    codetab: &mut [Codetab],
    queue: &mut CodeQueue,
) -> io::Result<u16> {
    // assert(sizeof(code) * CHAR_BIT >= *code_size);
    let code = is.read::<u16>(*code_size as u32)?;

    // Handle regular codes (the common case).
    if code != CONTROL_CODE as u16 {
        return Ok(code);
    }

    // Handle control codes.
    let control_code = is.read::<u16>(*code_size as u32)?;
    if control_code == INC_CODE_SIZE {
        if *code_size >= MAX_CODE_SIZE {
            return Err(io::Error::new(ErrorKind::InvalidData, "tried to increase code size when already at maximum));
        }
        (*code_size) += 1;
        return read_code(is, code_size, codetab, queue);
    }
    if control_code == PARTIAL_CLEAR {
        unshrink_partial_clear(codetab, queue);
        return read_code(is, code_size, codetab, queue);
    }
    Err(io::Error::new(ErrorKind::InvalidData, format!("Invalid control code {}", control_code)))
}
```